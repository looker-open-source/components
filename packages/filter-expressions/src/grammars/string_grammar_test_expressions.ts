/*

 MIT License

 Copyright (c) 2022 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import type { GrammarTestItem } from './grammar_test_utils'

export const stringGrammarTestItems: GrammarTestItem[] = [
  { expression: 'FOO', describe: 'is FOO', type: 'match', output: 'FOO' },
  {
    expression: 'FOO,BAR',
    describe: 'is FOO or BAR',
    type: 'match',
    output: 'FOO,BAR',
  },
  {
    expression: '%FOO%',
    describe: 'contains FOO',
    type: 'contains',
    output: '%FOO%',
  },
  {
    expression: '%FOO%',
    describe: 'contains FOO',
    type: 'contains',
    output: '%FOO%',
  },
  {
    expression: '%100^%^_^^ FOO 100^%^_^^%',
    describe: 'contains 100%_^ FOO 100%_^',
    type: 'contains',
    output: '%100^%^_^^ FOO 100^%^_^^%',
  },
  {
    expression: 'FOO%',
    describe: 'starts with FOO',
    type: 'startsWith',
    output: 'FOO%',
  },
  {
    expression: '100^%^_^^ FOO%',
    describe: 'starts with 100%_^ FOO',
    type: 'startsWith',
    output: '100^%^_^^ FOO%',
  },
  {
    expression: '%FOO',
    describe: 'ends with FOO',
    type: 'endsWith',
    output: '%FOO',
  },
  {
    expression: '%FOO 100^%^_^^',
    describe: 'ends with FOO 100%_^',
    type: 'endsWith',
    output: '%FOO 100^%^_^^',
  },
  {
    expression: 'EMPTY',
    describe: 'is blank',
    type: 'blank',
    output: 'EMPTY',
  },
  {
    expression: 'empty',
    describe: 'is blank',
    type: 'blank',
    output: 'EMPTY',
  },
  { expression: 'NULL', describe: 'is null', type: 'null', output: 'NULL' },
  { expression: 'null', describe: 'is null', type: 'null', output: 'NULL' },
  {
    expression: '-FOO',
    describe: 'is not FOO',
    type: '!match',
    output: '-FOO',
  },
  {
    expression: '-FOO,-BAR',
    describe: 'is not FOO or BAR',
    type: '!match',
    output: '-FOO,-BAR',
  },
  {
    expression: '-%FOO%',
    describe: 'does not contain FOO',
    type: '!contains',
    output: '-%FOO%',
  },
  {
    expression: '-FOO%',
    describe: 'does not start with FOO',
    type: '!startsWith',
    output: '-FOO%',
  },
  {
    expression: '-%FOO',
    describe: 'does not end with FOO',
    type: '!endsWith',
    output: '-%FOO',
  },
  {
    expression: '-EMPTY',
    describe: 'is not blank',
    type: '!blank',
    output: '-EMPTY',
  },
  {
    expression: '-empty',
    describe: 'is not blank',
    type: '!blank',
    output: '-EMPTY',
  },
  {
    expression: '-null',
    describe: 'is not null',
    type: '!null',
    output: '-NULL',
  },
  {
    expression: 'FOO%,BAR',
    describe: 'starts with FOO or is BAR',
    type: 'startsWith',
    output: 'FOO%,BAR',
  },
  {
    expression: 'FOO%,-FOOD',
    describe: 'starts with FOO, and is not FOOD',
    type: 'startsWith',
    output: 'FOO%,-FOOD',
  },
  {
    expression: 'F%OD',
    describe: 'F%OD',
    type: 'other',
    output: 'F%OD',
  },
  {
    expression: '_UF',
    describe: '_UF',
    type: 'other',
    output: '_UF',
  },
  // We should keep leading and trailing spaces
  {
    expression: '  hello  ',
    describe: 'is   hello  ',
    type: 'match',
    output: '^ ^ hello^ ^ ^ ',
  },
  {
    expression: '',
    describe: 'is any value',
    type: 'match',
    output: '',
  },
  /// Special CASES that should be typed as MATCH
  // Unquoted string containing the '-' character
  {
    expression: 't-shirt',
    describe: 'is t-shirt',
    type: 'match',
    output: 't-shirt',
  },
  // Quoted string containing the '-' character
  {
    expression: '"t-shirt \\"new\\" style"',
    describe: 'is "t-shirt "new" style"',
    type: 'match',
    output: '"t-shirt \\"new\\" style"',
  },
  // Quoted string containing the '%' character
  {
    expression:
      '"100% Silk Camisole \\"Angelina Jolie\'s Favorite\\" By Mary Green"',
    describe:
      'is "100% Silk Camisole "Angelina Jolie\'s Favorite" By Mary Green"',
    type: 'match',
    output:
      '"100% Silk Camisole \\"Angelina Jolie\'s Favorite\\" By Mary Green"',
  },
  // Quoted string containing the ',' character
  {
    expression: '"\\"DANCE\\" Tube Socks,T1418,multi-co..."',
    describe: 'is ""DANCE" Tube Socks,T1418,multi-co..."',
    type: 'match',
    output: '"\\"DANCE\\" Tube Socks,T1418,multi-co..."',
  },
  // Quoted string containing the '^' character
  {
    expression: '"Intentional^caret"',
    describe: 'is Intentional^caret',
    type: 'match',
    output: '"Intentional^caret"',
  },
  // User attribute example: there is no description since we have no value at that point
  {
    expression: `{{ _user_attributes['id'] }}`,
    describe: ``,
    type: 'user_attribute',
    output: `{{ _user_attributes['id'] }}`,
  },
  {
    expression: 'sub^_region',
    describe: 'is sub_region',
    type: 'match',
    output: '"sub_region"',
  },
  {
    expression: '"quoted_underscore"',
    describe: 'is quoted_underscore',
    type: 'match',
    output: '"quoted_underscore"',
  },
  {
    expression: 'hello^ ^ ',
    describe: 'is hello ',
    type: 'match',
    output: 'hello^ ^ ',
  },
  {
    expression: '^ whitespace^ ^ ',
    describe: 'is  whitespace ',
    type: 'match',
    output: '^ whitespace^ ^ ',
  },
  {
    expression: '%^ whitespace^ %',
    describe: 'contains whitespace',
    type: 'contains',
    output: '%whitespace%',
  },
]
