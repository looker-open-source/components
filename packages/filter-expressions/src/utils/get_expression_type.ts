/*

 MIT License

 Copyright (c) 2022 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import { Category } from '@looker/sdk'
import type { IDashboardFilter, ILookmlModelExploreField } from '@looker/sdk'
import type { FilterExpressionType } from '../types/filter_type'

interface FilterExpressionTypeMap {
  [type: string]: FilterExpressionType
}

const getTimeframeExpressionType = (fieldType?: string) => {
  if (!fieldType) return 'date'

  const isDateTime = ['date_time', 'hour', 'minute', 'second'].some(
    (timeString: string) => {
      return fieldType.indexOf(timeString) > -1
    }
  )
  if (isDateTime) return 'date_time'

  return 'date'
}

/**
 * yields a value for the 'expressionType' prop on the Filter component
 */
export const getExpressionTypeFromField = (
  field: ILookmlModelExploreField
): FilterExpressionType => {
  if (field?.category === Category.parameter && field?.type === 'number') {
    // If parameter field has enumerations it always parsed as 'tier' but
    // visually should be treated differently if number type defined in LookML
    // bugs: b/187940941, b/199507872
    return field.type as FilterExpressionType
  }
  if (field.enumerations) {
    return 'tier'
  }
  if (field.is_numeric) {
    return 'number'
  }
  if (field.is_timeframe) {
    return getTimeframeExpressionType(field.type)
  }
  if (field.type === 'location' || field.type === 'location_bin_level') {
    return 'location'
  }
  return 'string'
}

/**
 * Returns a valid filter expression type from a DashboardFilter object
 * depending if the filter is of type field or not
 */
export const getExpressionType = (
  filter: Pick<IDashboardFilter, 'type' | 'field'>
): FilterExpressionType => {
  if (filter.field) {
    return getExpressionTypeFromField(filter.field as ILookmlModelExploreField)
  }

  const filterExpressionMap: FilterExpressionTypeMap = {
    number_filter: 'number',
    string_filter: 'string',
    date_filter: 'date',
    // Catch for broken cases where field is not defined, display as a
    // string filter
    field_filter: 'string',
  }
  return filterExpressionMap[filter.type!]
}
