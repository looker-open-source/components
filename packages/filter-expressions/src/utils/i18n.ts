/*

 MIT License

 Copyright (c) 2022 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import { i18nInit as i18nInitBase } from '@looker/i18n'
import type { I18nState as _I18nState } from '@looker/i18n'
import merge from 'lodash/merge'
import { en } from '../locales'

export type I18nState = _I18nState
/**
 * Directly initialize the localization instance
 */
export async function i18nInit(options: I18nState = en) {
  const resources = merge(options.resources, en.resources)
  return i18nInitBase({ ...options, resources })
}
