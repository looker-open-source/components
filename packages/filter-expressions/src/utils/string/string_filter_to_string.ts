/*

 MIT License

 Copyright (c) 2022 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import isEmpty from 'lodash/isEmpty'
import type {
  FilterASTNode,
  FilterItemToStringFunction,
  FilterItemToStringMapType,
  FilterModel,
} from '../../types'
import isItemToString from '../to_string/is_item_to_string'
import { treeToString } from '../tree/tree_to_string'
import { userAttributeToString } from '../user_attribute/user_attribute_to_string'
import { escapeLeadingAndTrailingWhitespaces } from './escape_leading_and_trailing_whitespaces'
import { escapeWithCaret } from './escape_with_caret'
import { quoteFilter } from './quote_filter'

/**
 * stringFilterToString.ts
 * A collection of util functions that convert string grammar FilterItemProps (match, contains, etc)
 * to their filter_expression representation.
 * A FilterItem of:           will convert to:
 * {
 *   "is": true,              "FOO, BAR"
 *   "type": "match",
 *   "value": Array [
 *     FOO, BAR
 *   ],
 * }
 */

/**
 * Will double escape trailing spaces (used for match and endsWith filter types)
 */
const escapeWithDoubleLastEscape = (v: string) =>
  escapeLeadingAndTrailingWhitespaces(v)
/**
 * Will not double escape trailing spaces (used for startsWith and contains filter types
 * because a % sign will be the real end of the string)
 */
const escapeWithoutDoubleLastEscape = (v: string) =>
  escapeLeadingAndTrailingWhitespaces(v, false)

const matchToString = ({ value, is }: FilterModel) =>
  isItemToString(is, '', '-') +
  value
    .map(quoteFilter)
    .map(escapeWithDoubleLastEscape)
    .join(`,${isItemToString(is, '', '-')}`)

const multiValueToString = (
  values: string[],
  toString: (token: string) => string
) => values.map(toString).join(',')

const startWithToString = ({ value, is }: FilterModel) =>
  multiValueToString(
    value.map(escapeWithCaret).map(escapeWithoutDoubleLastEscape),
    (token: string) => `${isItemToString(is, '', '-') + String(token)}%`
  )

const endsWithToString = ({ value, is }: FilterModel) =>
  multiValueToString(
    value.map(escapeWithCaret).map(escapeWithDoubleLastEscape),
    (token: string) => `${isItemToString(is, '', '-')}%${String(token)}`
  )

const containsToString = ({ value, is }: FilterModel) =>
  multiValueToString(
    value.map(escapeWithCaret).map(escapeWithoutDoubleLastEscape),
    (token: string) => `${isItemToString(is, '', '-')}%${String(token)}%`
  )

const otherToString = ({ value, is }: FilterModel) =>
  multiValueToString(
    value,
    (token: string) => `${isItemToString(is, '', '-')}${String(token)}`
  )

const blankToString = ({ is }: FilterModel) =>
  `${isItemToString(is, '', '-')}EMPTY`

const nullToString = ({ is }: FilterModel) =>
  `${isItemToString(is, '', '-')}NULL`

const anyvalueToString = () => ''

const filterToStringMap: FilterItemToStringMapType = {
  startsWith: startWithToString,
  endsWith: endsWithToString,
  contains: containsToString,
  match: matchToString,
  blank: blankToString,
  null: nullToString,
  user_attribute: userAttributeToString,
  anyvalue: anyvalueToString,
  other: otherToString, // doesn't match to a UI component
}

/**
 * Maps a FilterItem to a function for converting it to an expression
 */
const stringToExpression = (item: FilterModel): string => {
  const toStringFunction: FilterItemToStringFunction =
    filterToStringMap[item.type]
  return toStringFunction?.(item) || ''
}

/**
 * itemIsNotEmpty is to filter out empty nodes when
 * building the string filter expression
 */
const itemIsNotEmpty = ({ type, value }: FilterModel) =>
  !(
    ['match', 'contains', 'startsWith', 'endsWith', 'other'].indexOf(type) >
      -1 && isEmpty(value)
  )
/**
 * Converts the AST to an array of FilterItems and then
 * converts each item into its expression representation
 */
export const stringFilterToString = (root: FilterASTNode) =>
  treeToString(root, stringToExpression, itemIsNotEmpty)
