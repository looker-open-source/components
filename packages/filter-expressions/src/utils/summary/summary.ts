/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import type { ILookmlModelExploreField } from '@looker/sdk'
import type {
  FilterASTNode,
  FilterItemToStringFunction,
  FilterModel,
  FilterExpressionType,
  UserAttributeWithValue,
} from '../../types'
import { hasMatchesAdvancedNode } from '../has_matches_advanced_node'
import { hasUserAttributeNodeWithoutValue } from '../has_user_attribute_node_without_value'
import { parseFilterExpression } from '../parse_filter_expression'
import { inorderTraversal } from '../tree'
import { typeToGrammar } from '../type_to_grammar'
import i18next from 'i18next'
const t = i18next.t.bind(i18next)

/**
 * Traverses ast and calls the describe function for every node
 * combines results using or / and labels
 */
const treeToSummary = (
  root: FilterASTNode,
  describe: FilterItemToStringFunction,
  filterType: FilterExpressionType,
  field?: ILookmlModelExploreField | null
): string => {
  const orItems: string[] = []
  const andItems: string[] = []
  inorderTraversal(root, (node: FilterASTNode) => {
    const item = node as FilterModel
    if (item.type !== ',') {
      ;(item.is ? orItems : andItems).push(describe(item, filterType, field))
    }
  })
  const resultOr = orItems ? orItems.join(' or ') : ''
  const resultAnd = andItems ? andItems.join(' and ') : ''
  let result = resultOr
  result += resultOr && resultAnd ? ', and ' : ''
  result += resultAnd

  return result
}

interface ISummaryOptions {
  /** The type of filter expression - number, string, etc. */
  type: FilterExpressionType
  /** The current value being filtered */
  expression?: string
  /** User attributes to be considered */
  userAttributes?: UserAttributeWithValue[]
  /** Field being filtered */
  field?: ILookmlModelExploreField | null
  /** Filter is required if true; required fields must have a value */
  required?: boolean
}

type ISummary = (o: ISummaryOptions) => string

/**
 * Builds a summary description for a filter expression
 */
export const summary: ISummary = (props) => {
  const {
    type,
    expression = '',
    userAttributes,
    field,
    required,
  } = props as ISummaryOptions

  if (required && !expression) {
    return t('Value required', { ns: 'summary' })
  }

  const { describe, subTypes } = typeToGrammar(type)
  const ast = parseFilterExpression(type, expression, userAttributes)

  // Special case: user attribute should be displayed
  // as the name and a "(null)" text if it has no value
  if (hasUserAttributeNodeWithoutValue(ast)) {
    const userAttribute = getUserAttributeMatchingAST(ast, userAttributes)
    if (userAttribute) {
      return `${userAttribute!.label} (null)`
    }
  }

  const isMatchesAdvanced = hasMatchesAdvancedNode(subTypes)(ast)
  return isMatchesAdvanced
    ? expression
    : treeToSummary(ast, describe, type, field)
}

const getUserAttributeMatchingAST = (
  { attributeName }: FilterASTNode,
  userAttributes?: UserAttributeWithValue[]
) => userAttributes?.find((ua) => ua.name === attributeName)
