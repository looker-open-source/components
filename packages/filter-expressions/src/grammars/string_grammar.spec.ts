/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import { parseFilterExpression } from '../utils'
import type { GrammarTestItem } from './grammar_test_utils'
import { stringGrammarTestItems } from './string_grammar_test_expressions'

import { convertTypeToOption } from '../utils/option/convert_type_to_option'
import { stringFilterToString } from '../utils/string/string_filter_to_string'
import { summary } from '../utils/summary/summary'
import { treeToList } from '../utils/tree/tree_to_list'

const filterType = 'string'
const testStringItem = (testItem: GrammarTestItem) => {
  test(`${testItem.expression}`, () => {
    const { expression, type, describe, output } = testItem
    const ast = parseFilterExpression(filterType, expression)
    const description = summary({ type: filterType, expression })
    expect(ast).toMatchSnapshot()
    const list = treeToList(ast)
    const item = list[0]

    const itemType =
      type === 'matchesAdvanced' ? item.type : convertTypeToOption(item)
    expect(itemType).toEqual(type)

    // test descriptions
    expect(description).toBe(describe)

    // test output
    const stringOutput =
      type === 'matchesAdvanced' ? expression : stringFilterToString(ast)
    expect(stringOutput).toBe(output)
  })
}

describe('String grammar can parse', () => {
  stringGrammarTestItems.forEach(testStringItem)
})

describe('String grammar can parse expanded character sets', () => {
  it('can parse japanese characters', () => {
    const expression = 'りんご,グレープ'
    const ast = parseFilterExpression('string', expression)
    const { type, value } = ast
    expect(type).toBe('match')
    expect(value).toHaveLength(2)
    expect(value[0]).toBe('りんご')
    expect(value[1]).toBe('グレープ')
    expect(ast).toMatchInlineSnapshot(`
      Object {
        "id": 1,
        "is": true,
        "left": undefined,
        "right": undefined,
        "type": "match",
        "value": Array [
          "りんご",
          "グレープ",
        ],
      }
    `)
  })
})
