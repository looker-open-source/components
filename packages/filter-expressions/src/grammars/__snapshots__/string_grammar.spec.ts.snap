// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`String grammar can parse   hello   1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": Array [
    "  hello  ",
  ],
}
`;

exports[`String grammar can parse  1`] = `
Object {
  "id": 1,
  "is": true,
  "type": "match",
  "value": Array [],
}
`;

exports[`String grammar can parse "\\"DANCE\\" Tube Socks,T1418,multi-co..." 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": Array [
    "\\"DANCE\\" Tube Socks,T1418,multi-co...",
  ],
}
`;

exports[`String grammar can parse "100% Silk Camisole \\"Angelina Jolie's Favorite\\" By Mary Green" 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": Array [
    "100% Silk Camisole \\"Angelina Jolie's Favorite\\" By Mary Green",
  ],
}
`;

exports[`String grammar can parse "Intentional^caret" 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": Array [
    "Intentional^caret",
  ],
}
`;

exports[`String grammar can parse "quoted_underscore" 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": Array [
    "quoted_underscore",
  ],
}
`;

exports[`String grammar can parse "t-shirt \\"new\\" style" 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": Array [
    "t-shirt \\"new\\" style",
  ],
}
`;

exports[`String grammar can parse %^ whitespace^ % 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "contains",
  "value": Array [
    "whitespace",
  ],
}
`;

exports[`String grammar can parse %100^%^_^^ FOO 100^%^_^^% 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "contains",
  "value": Array [
    "100%_^ FOO 100%_^",
  ],
}
`;

exports[`String grammar can parse %FOO 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "endsWith",
  "value": Array [
    "FOO",
  ],
}
`;

exports[`String grammar can parse %FOO 100^%^_^^ 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "endsWith",
  "value": Array [
    "FOO 100%_^",
  ],
}
`;

exports[`String grammar can parse %FOO% 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "contains",
  "value": Array [
    "FOO",
  ],
}
`;

exports[`String grammar can parse %FOO% 2`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "contains",
  "value": Array [
    "FOO",
  ],
}
`;

exports[`String grammar can parse ^ whitespace^ ^  1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": Array [
    " whitespace ",
  ],
}
`;

exports[`String grammar can parse _UF 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "other",
  "value": Array [
    "_UF",
  ],
}
`;

exports[`String grammar can parse {{ _user_attributes['id'] }} 1`] = `
Object {
  "attributeName": "id",
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "user_attribute",
}
`;

exports[`String grammar can parse -%FOO 1`] = `
Object {
  "id": 1,
  "is": false,
  "left": undefined,
  "right": undefined,
  "type": "endsWith",
  "value": Array [
    "FOO",
  ],
}
`;

exports[`String grammar can parse -%FOO% 1`] = `
Object {
  "id": 1,
  "is": false,
  "left": undefined,
  "right": undefined,
  "type": "contains",
  "value": Array [
    "FOO",
  ],
}
`;

exports[`String grammar can parse -EMPTY 1`] = `
Object {
  "id": 1,
  "is": false,
  "left": undefined,
  "right": undefined,
  "type": "blank",
}
`;

exports[`String grammar can parse -FOO 1`] = `
Object {
  "id": 1,
  "is": false,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": Array [
    "FOO",
  ],
}
`;

exports[`String grammar can parse -FOO% 1`] = `
Object {
  "id": 1,
  "is": false,
  "left": undefined,
  "right": undefined,
  "type": "startsWith",
  "value": Array [
    "FOO",
  ],
}
`;

exports[`String grammar can parse -FOO,-BAR 1`] = `
Object {
  "id": 1,
  "is": false,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": Array [
    "FOO",
    "BAR",
  ],
}
`;

exports[`String grammar can parse -empty 1`] = `
Object {
  "id": 1,
  "is": false,
  "left": undefined,
  "right": undefined,
  "type": "blank",
}
`;

exports[`String grammar can parse -null 1`] = `
Object {
  "id": 1,
  "is": false,
  "left": undefined,
  "right": undefined,
  "type": "null",
}
`;

exports[`String grammar can parse 100^%^_^^ FOO% 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "startsWith",
  "value": Array [
    "100%_^ FOO",
  ],
}
`;

exports[`String grammar can parse EMPTY 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "blank",
}
`;

exports[`String grammar can parse F%OD 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "other",
  "value": Array [
    "F%OD",
  ],
}
`;

exports[`String grammar can parse FOO 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": Array [
    "FOO",
  ],
}
`;

exports[`String grammar can parse FOO% 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "startsWith",
  "value": Array [
    "FOO",
  ],
}
`;

exports[`String grammar can parse FOO%,-FOOD 1`] = `
Object {
  "id": 2,
  "left": Object {
    "id": 1,
    "is": true,
    "left": undefined,
    "right": undefined,
    "type": "startsWith",
    "value": Array [
      "FOO",
    ],
  },
  "right": Object {
    "id": 3,
    "is": false,
    "left": undefined,
    "right": undefined,
    "type": "match",
    "value": Array [
      "FOOD",
    ],
  },
  "type": ",",
}
`;

exports[`String grammar can parse FOO%,BAR 1`] = `
Object {
  "id": 2,
  "left": Object {
    "id": 1,
    "is": true,
    "left": undefined,
    "right": undefined,
    "type": "startsWith",
    "value": Array [
      "FOO",
    ],
  },
  "right": Object {
    "id": 3,
    "is": true,
    "left": undefined,
    "right": undefined,
    "type": "match",
    "value": Array [
      "BAR",
    ],
  },
  "type": ",",
}
`;

exports[`String grammar can parse FOO,BAR 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": Array [
    "FOO",
    "BAR",
  ],
}
`;

exports[`String grammar can parse NULL 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "null",
}
`;

exports[`String grammar can parse empty 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "blank",
}
`;

exports[`String grammar can parse hello^ ^  1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": Array [
    "hello ",
  ],
}
`;

exports[`String grammar can parse null 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "null",
}
`;

exports[`String grammar can parse sub^_region 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": Array [
    "sub_region",
  ],
}
`;

exports[`String grammar can parse t-shirt 1`] = `
Object {
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "match",
  "value": Array [
    "t-shirt",
  ],
}
`;
